File: antilles-portal/src/service/alarm.js
/*
 * Copyright © 2019-present Lenovo
 * 
 * This file is licensed under both the BSD-3 license for individual use and
 * EPL-1.0 license for commercial use. Full text of both licenses can be found in
 * COPYING.BSD and COPYING.EPL files.
 */

import Vue from 'vue'
import Parser from '../common/parser'
import ErrorHandler from '../common/error-handler'
import TableDataFetcherFactory from '../common/table-data-fetcher-factory'

const AlarmStatusEnums = ['present', 'confirmed', 'resolved'];

class Alarm {
  constructor() {
    this.id = 0;
    this.policyId = '';
    this.policyName = '';
    this.policyLevel = '';
    this.status = '';
    this.createTime = new Date(0);
    this.nodeId = 0;
    this.nodeName = '';
    this.gpuId = '';
    this.comment = '';
  }
  static parseFromRestApi(jsonObj) {
    var alarm = new Alarm();
    alarm.id = jsonObj.id;
    alarm.policyId = jsonObj.policy__id;
    alarm.policyName = jsonObj.policy__name;
    alarm.policyLevel = jsonObj.policy__level.toLowerCase();
    alarm.status = jsonObj.status.toLowerCase();
    alarm.createTime = Parser.parseTimeFromRestApi(jsonObj.create_time);
    alarm.nodeId = 0;
    alarm.gpuId = jsonObj.gpu_index;
    alarm.nodeName = jsonObj.node;
    alarm.comment = jsonObj.comment;
    return alarm;
  }
  get id() {
    return this._id;
  }
  set id(id) {
   this._id = id;
  }
  get id() {
    return this._id;
  }
  set id(id) {
   this._id = id;
  }
  get policyId() {
    return this._policyId;
  }
  set policyId(policyId) {
   this._policyId = policyId;
  }
  get policyName() {
    return this._policyName;
  }
  set policyName(policyName) {
   this._policyName = policyName;
  }
  get policyLevel() {
    return this._policyLevel;
  }
  set policyLevel(policyLevel) {
   this._policyLevel = policyLevel;
  }
  get status() {
    return this._status;
  }
  set status(status) {
   this._status = status;
  }
  get createTime() {
    return this._createTime;
  }
  set createTime(createTime) {
   this._createTime = createTime;
  }
  get nodeId() {
    return this._nodeId;
  }
  set nodeId(nodeId) {
   this._nodeId = nodeId;
  }
  get nodeName() {
    return this._nodeName;
  }
  set nodeName(nodeName) {
   this._nodeName = nodeName;
  }
  get gpuId() {
    return this._gpuId;
  }
  set gpuId(gpuId) {
   this._gpuId = gpuId;
  }
  get comment() {
    return this._comment;
  }
  set comment(comment) {
   this._comment = comment;
  }
}

class AlarmPrompt {
  constructor() {
    this.count = 0;
    this.isSound = Boolean;
  }

  static parseFromRestApi(jsonObj) {
    var alarmPrompt = new AlarmPrompt();
    alarmPrompt.count = jsonObj.count;
    alarmPrompt.isSound = jsonObj.sound;
    return alarmPrompt;
  }

  get count() {
    return this._count;
  }
  set count(count) {
   this._count = count;
  }
  get isSound() {
    return this._isSound;
  }
  set isSound(isSound) {
   this._isSound = isSound;
  }

}
function buildFilters(idList, statusList, policyLevelList, timeRange) {
  var filters = [];
  if(idList) {
    filters.push({prop: "id", type: "in", values: idList});
  }
  if(statusList) {
    filters.push({prop: "status", type: "in", values: statusList});
  }
  if(policyLevelList) {
    filters.push({prop: "policy__level", type: "in", values: policyLevelList});
  }
  if(timeRange) {
    filters.push({prop: "create_time", type: "range", values: timeRange});
  }
  return filters
}

function processByFilters(filters, action) {
  return new Promise((resolve, reject) => {
    var req = {
      filters: filters,
      action: action
    };
    Vue.http.post('/api/alarm/status', req).then((res) => {
        resolve(res.body);
      },
      (res) => {
        ErrorHandler.restApiErrorHandler(res, reject);
      }
    );
  });
}


function alarmsTableDataParser(res) {
  var alarms = [];
  res.data.forEach((item) => {
    alarms.push(Alarm.parseFromRestApi(item));
  });
  return {
    offset: res.offset,
    total: res.total,
    data: alarms
  };
}

function alarmsRestApiPropMap(prop) {
  if (prop == 'id')
    return 'id';
  else if (prop == 'policyId')
    return 'policy__id';
  else if (prop == 'policyName')
    return 'policy__name';
  else if (prop == 'policyLevel')
    return 'policy__level';
  else if (prop == 'status')
    return 'status';
  else if (prop == 'createTime')
    return 'create_time';
  else if (prop == 'nodeName')
    return 'node';
  else if (prop == 'gpuId')
    return 'gpu_index';
  else if (prop == 'comment')
    return 'comment';
  else if (prop == 'nodeName')
    return 'node';
  else
    return '';
}


function getAlarmTableDataFetcher() {
  return TableDataFetcherFactory.createRemotePagingFetcher('/api/alarm/alarm', alarmsRestApiPropMap, alarmsTableDataParser, 'data', 'offset', 'total');
}
function confirmAlarms(idList, statusList, policyLevelList, timeRange) {
  return processByFilters(buildFilters(idList, statusList, policyLevelList, timeRange), 'confirm');
}
function fixAlarms(idList, statusList, policyLevelList, timeRange) {
  return processByFilters(buildFilters(idList, statusList, policyLevelList, timeRange), 'solve');
}
function deleteAlarms(idList, statusList, policyLevelList, timeRange) {
  return processByFilters(buildFilters(idList, statusList, policyLevelList, timeRange), 'delete');
}


function updateAlarmComment(id, comment) {
  return new Promise((resolve, reject) => {
    var req = {
      comment: comment
    };
    Vue.http.post('/api/alarm/'+id+'/comment/', req).then((res) => {

        resolve(res.body);
      },
      (res) => {
        ErrorHandler.restApiErrorHandler(res, reject);
      }
    );
  });
}

function getAlarmSound() {
  return new Promise((resolve, reject) => {
    Vue.http.get('api/alarm/sound/').then((res) => {
      resolve(AlarmPrompt.parseFromRestApi(res.body))
    }, (res) => {
      ErrorHandler.restApiErrorHandler(res, reject);
    })
  });
}

export default {
  AlarmStatusEnums,
  getAlarmTableDataFetcher,
  confirmAlarms,
  fixAlarms,
  deleteAlarms,
  updateAlarmComment,
  getAlarmSound
}

File: antilles-core/openHPC_web_project/antilles/alarm/views/alarm.py
# -*- coding: utf-8 -*-

"""
Copyright © 2019-present Lenovo

This file is licensed under both the BSD-3 license for individual use and
EPL-1.0 license for commercial use. Full text of both licenses can be found in
COPYING.BSD and COPYING.EPL files.
"""

import logging
import os
from datetime import datetime

from django.conf import settings
from django.utils.dateparse import parse_datetime
from django.utils.timezone import localtime
from rest_framework.response import Response
from rest_framework.status import HTTP_200_OK, HTTP_204_NO_CONTENT
from rest_framework.views import APIView

from antilles.common.utils import json_schema_validate
from antilles.common.views import DataTableView
from antilles.optlog.optlog import EventLog
from antilles.user.permissions import AsAdminRole, AsOperatorRole

from ..models import Alarm, Policy

logger = logging.getLogger(__name__)


class AlarmView(DataTableView):
    permission_classes = (AsOperatorRole,)

    def get_query(self, request, *args, **kwargs):
        return Alarm.objects

    def filters_params(self, args, prop, replace_func):
        """
        Use replace_func to replace element in filters[index]['values']
        """
        index = None
        for i, f in enumerate(args['filters']):
            if f['prop'] == prop:
                index = i
                break
        if index:
            args['filters'][index]['values'] = \
                [replace_func(v) for v in args['filters'][index]['values']]
        return args

    def replace_policy_level(self, e):
        values = {
            'fatal': Policy.FATAL,
            'warn': Policy.WARN,
            'error': Policy.ERROR,
            'info': Policy.INFO,
        }
        return values[e]

    def replace_create_time(self, e):
        return parse_datetime(e)

    def params(self, args):
        args = self.filters_params(args, 'policy__level',
                                   self.replace_policy_level)
        args = self.filters_params(args, 'create_time',
                                   self.replace_create_time)
        return args

    def trans_result(self, result):
        return {
            'id': result.id,
            'policy__id': result.policy.id,
            'policy__name': result.policy.name,
            'policy__level': result.policy.get_level_display(),
            'status': result.status,
            'create_time': localtime(result.create_time),
            'node': result.node,
            'comment': result.comment,
            'index': result.index,
        }

    @json_schema_validate({
        "type": "object",
        "properties": {
            "filters": {
                "type": "array",
                "properties": {
                    "prop": {"type": "string"},
                    "type": {"type": "string"},
                    "values": {"type": "array"}
                }
            },
            "action": {
                "type": "string",
                "enum": ['confirm', 'solve', 'delete']
            }
        },
        "required": ["filters", "action"]
    })
    def post(self, request):
        params = self.params(request.data)

        index = None
        for i, f in enumerate(params['filters']):
            if f['prop'] == 'id':
                index = i
                break
        if not params['filters'][index]['values']:
            del params['filters'][index]

        query = self.get_query(request)
        query = self.filters(query, params['filters'])  # filter

        # add operationlog
        [EventLog.opt_create(
            request.user.username,
            EventLog.alarm,
            params['action'],
            EventLog.make_list(alarm.id, alarm.policy.name)
            ) for alarm in query]

        if params['action'] == 'confirm':
            query = query.filter(status__in=['present'])
            query.update(status=Alarm.CONFIRMED)
        elif params['action'] == 'solve':
            query.update(status=Alarm.RESOLVED)
        elif params['action'] == 'delete':
            query.delete()

        return Response({
            'ret': 'success',
        }, status=HTTP_200_OK)


class CommentView(APIView):
    permission_classes = (AsOperatorRole,)

    @json_schema_validate({
        "type": "object",
        "properties": {
            "comment": {"type": "string"}
        },
        "required": ["comment"]
    })
    def post(self, request, pk):
        comment = request.data.get('comment', None)
        alarm = Alarm.objects.get(id=pk)
        alarm.comment = comment
        alarm.save()

        # add operationlog
        EventLog.opt_create(
            request.user.username,
            EventLog.alarm,
            EventLog.comment,
            EventLog.make_list(pk, alarm.policy.name)
        )

        return Response(status=HTTP_204_NO_CONTENT)


class ScriptView(APIView):
    permission_classes = (AsAdminRole, )

    def get(self, request):
        scripts_info = [
            {
                'name': script[0],
                'size': os.path.getsize(script[1]),
                'modify_time': datetime.fromtimestamp(
                    os.path.getmtime(script[1])
                ).strftime('%Y-%m-%d %H:%M:%S')
            } for script in [
                (filename, os.path.join(settings.SCRIPTS_DIR, filename))
                for filename in os.listdir(settings.SCRIPTS_DIR)
                if not os.path.isdir(filename)
                and not filename.startswith('.')
            ]
        ]
        return Response(scripts_info)


class SoundView(APIView):
    permission_classes = (AsOperatorRole,)

    def get(self, request):
        query = Alarm.objects.filter(status=Alarm.PRESENT)
        return Response({
            'count': query.count(),
            'sound': query.filter(
                policy__in=Policy.objects.filter(sound=True)).exists()
        })
